Removed PerCellDamageWarhead

reaperrr opened the pull request:
The only reason why I haven't submitted this ages ago is that I have no idea how to tackle this with regard to an upgrade rule.

If this is supposed to have an upgrade rule, either I'll need help or someone else has to take over, unless a simple `PerCellDamage` -> `SpreadDamage` renaming to avoid crashes while ignoring `Size`/`Falloff` is considered sufficient.

reaperrr committed:
Remove PerCellDamageWarhead

Remove redundant CrateNuke definition from Training Camp

pchote commented on mods/ra/maps/training-camp/map.yaml:
```diff
@@ -1058,9 +1058,10 @@ VoxelSequences:
 			Damage: 250
-			Size: 4
+			Spread: 1c0
+			Falloff: 100, 100, 100, 100, 100, 0
```
Maybe port and rebalance the proper nuke warhead, instead of trying to recreate the old behaviour?

reaperrr replied:
Would be a cleaner solution, I'll take another look later.

abcdefg30 replied:
Is this solved?

abcdefg30 replied:
Where is this actually used (in this map)?

obrakmann replied:
You can capture the war factory in the middle and build a demo truck. I don't think we should put that much effort into fine-tuning the nuke warhead for this map. I doubt anyone ever plays this map. Like, at all.

abcdefg30 replied:
Demo trucks use the [```MiniNuke```](https://github.com/OpenRA/OpenRA/blob/e875240ca35d537d484a0ec16868c3b6145f6dc1/mods/ra/rules/vehicles.yaml#L653) instead of the ```CrateNuke```.

obrakmann replied:
Oh, right. Then... it doesn't seem to get used. Crates are disabled, and the CrateSpawner is removed. @reaperrr, just remove the whole CrateNuke definition here.

reaperrr replied:
Alright, will do that.

pchote commented:
This is a tricky case, so I'd be okay with outright removing the warhead and then logging a message that it needs to be fixed manually.

obrakmann commented:
Damage-wise, the nuke seems to behave as before.

So, the way I see it, these things are left to be done:

- [ ] Rebase
- [ ] Add an upgrade rule (similar to [this](https://github.com/OpenRA/OpenRA/commit/1f048b64ac819638597fe489dc20b923e370089d), as pchote mentioned)
- [ ] Remove the actual warhead code itself

reaperrr committed:
Upgrade rule for PerCellDamage removal

reaperrr commented:
Removed actual warhead code, rebased, added upgrade rule.

I left the yaml code as-is for now, since I don't have the time to properly balance test other approaches. The way it's now, It should do roughly the same as the old code and therefore not upset anything, balance-wise.

obrakmann commented:
:+1: Tested with a side-by-side comparison with the current playtest, there was no discernible difference in warhead damage.

reaperrr commented:
Updated, removed the redundant CrateNuke definition from Training Camp.

abcdefg30 commented:
I don't know what happened, but a demo truck equipped with ```CrateNuke``` (using this PR) did less damage then it does in the playtest/at bleed.

reaperrr commented:
> I don't know what happened, but a demo truck equipped with CrateNuke (using this PR) did less damage then it does in the playtest/at bleed.

Well, the old PerCellDamage warhead did full damage on every cell within the radius, while the SpreadDamage has a (steep) falloff. 

My idea is to bring the overall damage a little closer to the MiniNuke by 
- increasing the impact warhead ```Spread``` from 0.25 cells to a full cell, so the damage of the impact warhead spreads over 5 cells instead of only 1.25.
- applying the less steep falloff curve of the areanuke to the impact warhead as well.
- increasing the areanuke ```Damage``` by 20%.
- using the same ```Versus``` values as on the other nukes (this slightly increases damage versus ```none```, moderately increases damage vs. ```light``` and significantly increases damage vs. ```heavy``` armor, while halving damage vs. ```concrete```)

I'll give the changes a test run before I push them.

reaperrr commented:
@abcdefg30: I've pushed the changes mentioned above, except instead of messing with the impact warhead Falloff, I made the areanuke Falloff even less steep.

Since nuke crates probably don't occur too often, I'd prefer any balance fine-tuning (if necessary) to be left for a separate PR, unless the updated implementation is still a show-stopper for some reason.

pchote commented:
Looks good to me :+1:.  Tested the TS ion cannon and the RA Bomber John map (which seems to be the only place left where the CrateNuke weapon can actually be triggered) and found no issues.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/2d66332f45b68962eb96a45a6ac48bccd024e89f%5E...2d66332f45b68962eb96a45a6ac48bccd024e89f
